app/services/post_scraper.rb

```diff
@@ -1,88 +1,59 @@
-class PostScraper < Object
-  SANDBOX_ID = Rails.env.production? ? 3 : 5
-  BASE_ACCOUNTS = {
-    'alicornucopia'      => 'Alicorn',
-    'pythbox'            => 'Kappa',
-    'lintamande'         => 'lintamande',
-    'marrinikari'        => 'Marri',
-    'peterxy'            => 'Pedro',
-    'peterverse'         => 'Pedro',
-    'curiousdiscoverer'  => 'CuriousDiscoverer',
-    'aestrix'            => 'Aestrix',
-    'unbitwise'          => 'Unbitwise',
-    'erinflight'         => 'ErinFlight',
-    'andaisq'            => 'andaisq',
-    'rockeye-stonetoe'   => 'Rockeye',
-    'rockeye_stonetoe'   => 'Rockeye',
-    'maggie-of-the-owls' => 'MaggieoftheOwls',
-    'maggie_of_the_owls' => 'MaggieoftheOwls', # have both - and _ versions cause Dreamwidth supports both
-    'nemoconsequentiae'  => 'Nemo',
-    'armokgob'           => 'Armok',
-    'timepoof'           => 'Timepoof',
-  }
+class PostScraper < Generic::Service
+  SANDBOX_ID = 3
 
-  attr_accessor :url, :post, :html_doc
+  attr_reader :url
 
-  def initialize(url, board_id=nil, section_id=nil, status=nil, threaded_import=false, console_import=false, subject=nil)
-    @board_id = board_id || SANDBOX_ID
+  def initialize(url, board_id: SANDBOX_ID, section_id: nil, status: Post::STATUS_COMPLETE, threaded: false, console: false, subject: nil)
+    @board_id = board_id
     @section_id = section_id
-    @status = status || Post::STATUS_COMPLETE
-    url += (url.include?('?') ? '&' : '?') + 'style=site' unless url.include?('style=site')
-    url += '&view=flat' unless url.include?('view=flat') || threaded_import
-    @url = url
-    @console_import = console_import
-    @threaded_import = threaded_import # boolean
+    @status = status
+    @console_import = console
+    @threaded_import = threaded # boolean
     @subject = subject
+    @url = clean_url(url)
+    super()
   end
 
-  def scrape!
-    @html_doc = doc_from_url(@url)
+  def scrape(threads=nil)
+    html_doc = doc_from_url(@url)
 
     Post.transaction do
-      import_post_from_doc(@html_doc)
-      import_replies_from_doc(@html_doc)
-      links = page_links
-      links.each_with_index do |link, i|
-        logger.debug "Scraping '#{@post.subject}': page #{i+1}/#{links.count}"
-        doc = doc_from_url(link)
-        import_replies_from_doc(doc)
+      import_post_from_doc(html_doc)
+      if threads.present?
+        threads.each { |thread| evaluate_links(doc_from_url(thread)) }
+      else
+        evaluate_links(html_doc)
       end
       finalize_post_data
     end
+    return if @errors.present?
```

Comment:
I'm concerned that we might not roll back transactions if an error occurs, any more – we'll be populating `@errors`, but may not raise anything to the Post.transaction block.

I'm not sure how to resolve this, but I think this transaction might require exception flow. Alternatively, maybe we could move this `return if @errors.present?` just into the transaction block – then we'll still be potentially do lots of work after encountering an error, but at least we'll ensure database consistency.

Reply:
Any point where errors are added we should also spawn a rollback. I can add another interrupt in here just in case?

Reply:
Hmm. I think it'd be sensible to add it just in case, and so we can make changes in the future without having to remember this. Does that sound okay?

Reply:
I'm not sure how to make it be tested

Reply:
Try creating a post like normal, but then:

`allow(post_scraper).to receive(:finalize_post_data) { post_scraper.instance_variable_set(:@errors, ['an error!']) }`

And make sure it isn't persisted?

Reply:
(Alternatively we could use getters and setters for the errors, and in this block here set an error using the setter. That way it'd be slightly less brittle, but also more work to implement now.)